docs/language/informal/enhanced-type-promotion.md

```diff
@@ -0,0 +1,540 @@
+
+Without type promotion, the marked line would have a static error since Object
+does not have a `.length` getter. Type promotion notes that an is String test
+was applied to `o`, and promotes or "re-types" `o` to have type String inside
+the then clause of the if statement.
+
+The type promotion rules in Dart are useful, but very restricted. Some
+restriction is necessary because dataflow based type inference can get quite
+complex in some cases. But, in practice, the current rules are too restrictive.
+For example, this does not work:
+
+```dart
+int stringLength(Object o) {
+  if (o is! String) {
+    throw "Not a string.";
+  } else {
+    return o.length; // Error: Object does not have a "length" getter.
+  }
+}
+```
+
+The type promotion rules don't handle negated `is` checks and else clauses. This
+doesn't work either:
+
+```dart
+int stringLength(Object o) {
+  if (o is! String) throw "Not a string.";
+  return o.length; // Error: Object does not have a "length" getter.
+}
+```
+
+Since the first line will definitely throw if `o` is not a String, the only way
+the second line could execute is if `o` is a string. But type promotion fails to
+understand that.
+
+To get around these limitations, the analyzer used by IntelliJ and others also
+has "type propagation". This is more sophisticated than type promotion. But,
+since the language spec controls what errors and warnings are shown, type
+propagation is limited to be using for things like auto-complete suggestions.
+
+Maintaining two similar but different flow typing systems adds complexity to our
+tools and confuses users. ("Why are you showing me an error that a variable
+isn't a Foo, when auto-complete on that same variable shows me all of Foo's
+methods?")
+
+To remedy that, this proposal augments type promotion with smarter rules when
+variables can be promoted. That should avoid requiring annoying redundant casts
+in more places, and let us unify type promotion and propagation into a single
+coherent system. This proposal doesn't cover all possible extensions to type
+propagation, but it handles the simplest, most common cases.
+
+## How Type Promotion Works
+
+The language specifies type promotion out of three pieces:
+
+1.  **Certain expressions cause a fact to be known about a variable.** For
+    example, `o` is String deduces the fact "`o` must be a String".
+
+2.  **The scope where a fact is known to be true.** For example in:
+
+    ```dart
+    if (o is String) {
+      return o.length;
+    } else {
+      throw "Not a string.";
+    }
+    ```
+
+    The fact "`o` is a String" is known to be true inside the then clause of the
+    if statement, but not the else clause, and not after the end of the if
+    statement.
+
+3.  **Which variables can have their types promoted.** Some code may cause a
+    fact to no longer be true. In that case, we shouldn't promote it. A simple
+    example is:
+
+    ```dart
+    if (o is String) {
+      o = 123; // Not a string anymore!
```

Comment:
Sure.  Note though that I'm not really suggesting the same thing that you addressed in your comment above: I'm not suggesting promoting based on the assignment, I'm suggesting just allowing promotion (from the `is` test) even if there is an assignment, and then allowing the assignment to become an error if it changes the type.